traits/src/signature.rs

```diff
@@ -0,0 +1,253 @@
+//! This module provides a common [`SignConsts`] trait and helper macros for signature scheme implementations.
+//!
+//! Instead of a fully trait-based API for signature operations, this crate provides a trait and
+//! macros that can be used to implement signature APIs with a given shape.
+//!
+//! ### Defining useful constants with the [`SignConsts`] trait
+//!
+//! Structs that implement the [`SignConsts`] trait allow retrieving useful constants for that
+//! signature algorithm. These can be used as the lengths of new buffers representing the
+//! signing key, verification key, signature, or the randomness input to keygen functions.
+//!
+//! Example:
+//! ```rust
+//! use libcrux_traits::signature::SignConsts;
+//! use libcrux_signature::mldsa::ml_dsa_44::MlDsa44;
+//!
+//! // the length of the signing key in bytes.
+//! assert_eq!(MlDsa44::SIGNING_KEY_LEN, 2560);
+//!
+//! // the length of the verification key in bytes.
+//! assert_eq!(MlDsa44::VERIFICATION_KEY_LEN, 1312);
+//!
+//! // the length of the signature in bytes.
+//! assert_eq!(MlDsa44::SIGNATURE_LEN, 2420);
+//!
+//! // the length of the rand_keygen buffer in bytes.
+//! assert_eq!(MlDsa44::RAND_KEYGEN_LEN, 32);
+//! ```
+//!
+//! ### Implementing key-centric APIs
+//!
+//! The [`impl_key_centric_types!()`] macro can be used to conveniently define types
+//! used by key-centric signature APIs.
+//! - `SigningKey`, `SigningKeyRef`: owned and borrowed signing keys, respectively.
+//! - `VerificationKey`, `VerificationKeyRef`: owned and borrowed verification keys, respectively.
+//! - `Signature`: an owned signature.
+//! - `KeyPair`: an owned signature keypair (contains a `SigningKey` and a `VerificationKey`)
+//!
+//! Key-centric APIs can be defined on the above structs.
+//! - `SigningKey::sign()`: The first argument should be the `payload`, followed by other
+//! arguments. This API should return an owned `Signature`.
+//! - `SigningKeyRef::sign()`: The first argument should be the `payload`, followed by a
+//! `signature: &mut [u8]` representing the signature buffer to write into.
+//! - `VerificationKey::verify()`: The first argument should be a slice representing the payload,
+//! followed by a reference to a `Signature`.
+//! - `VerificationKeyRef::verify()`: The first argument should be a slice representing the
+//! payload, followed by a `signature: &[u8]` representing the signature.
+//!
+//!  ```rust
+//! use libcrux_signature::ed25519::{Ed25519, KeyPair, SigningKey, VerificationKey};
+//! use libcrux_traits::signature::SignConsts;
+//!
+//! // generate a new signature keypair
+//! use rand::TryRngCore;
+//! let mut rng = rand::rngs::OsRng;
```

Comment:
Using rng's in this way suggests that we think its' a good idea. We shouldn't do that and at least add a comment about using proper rngs.